vllm/model_executor/layers/fused_moe/fused_moe.py

```diff
@@ -1174,6 +1176,79 @@ def fused_experts(hidden_states: torch.Tensor,
             block_shape=block_shape)
 
 
+# This is a new triton implementation of the fused_experts function
+# Note: this only supports outplace and automatically renormalizes
+# TODO: add support for more arguments (see fused_experts)
+# TODO: integrate with fused_moe outer function
+# TODO: check whether chunking is needed  (see fused_experts_impl)
+def fused_experts_triton_exp(hidden_states: torch.Tensor,
+                            w1: torch.Tensor,
+                            w2: torch.Tensor,
+                            router_logits: torch.Tensor,
+                            topk: int,
+                            activation: str = "silu") -> torch.Tensor:
+
+    # quantization not yet supported in new triton MoE kernel
+    assert hidden_states.dtype == torch.bfloat16, "hidden_states must be bfloat16"
+    assert w1.dtype == torch.bfloat16, "w1 must be bfloat16"
+    assert w2.dtype == torch.bfloat16, "w2 must be bfloat16"
+
+    # Shape check
+    assert hidden_states.ndim == 2, "hidden_states must be 2D"
+    assert hidden_states.shape[-1] == w1.shape[
+        -1], "hidden_states shape[-1] must be equal to w1 shape[-1]"
+    assert w1.shape[-1] == w2.shape[
+        1], "w1 shape[-1] must be equal to w2 shape[1]"
+
+
+
+    M, K = hidden_states.shape
+    N = w1.shape[1]
+
+    # compute new Triton style routing inputs
+    # we don't use scatter idx (why not?)
+    routing_data, gather_idx, _ = routing(router_logits, topk)
+
+    n_expts_tot = routing_data.n_expts_tot
+    n_expts_act = routing_data.n_expts_act
+
+    w1 = w1.transpose(-2, -1).contiguous()
+    w2 = w2.transpose(-2, -1).contiguous()
+
+    # consistent with default implementation
+    intermediate_cache2 = torch.empty((M * n_expts_act, N // 2),
+                                      device="cuda",
+                                      dtype=torch.bfloat16)
+    out_hidden_states = torch.empty_like(hidden_states)
+
+    intermediate_cache1 = matmul_ogs(hidden_states, w1, None, routing_data,
+                                     gather_indx)
+
+    # TODO: replace vllm impls here with triton impls -- 
```

Comment:
as discussed, at least for silu the triton impl cuts the tensor in half differently so need to address that before porting.